arch/arm64/core/cortex_r/mpu/arm_mpu.c

```diff
@@ -0,0 +1,193 @@
+/*
+ * Copyright (c) 2017 Linaro Limited.
+ * Copyright (c) 2021 Arm Limited (or its affiliates). All rights reserved.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <device.h>
+#include <init.h>
+#include <kernel.h>
+#include <soc.h>
+#include <arch/arm64/cortex_r/mpu/arm_mpu.h>
+#include <linker/linker-defs.h>
+
+#define LOG_LEVEL CONFIG_MPU_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_DECLARE(mpu);
+
+#define MPUIR_EL1	S3_0_c0_c0_4
+#define PRSELR_EL1	S3_0_c6_c2_1
+#define PRBAR_EL1	S3_0_c6_c8_0
+#define PRLAR_EL1	S3_0_c6_c8_1
+
+/*
+ * AArch64 Memory Model Feature Register 0
+ * Provides information about the implemented memory model and memory
+ * management support in AArch64 state.
+ * See Arm Architecture Reference Manual Supplement
+ *  Armv8, for Armv8-R AArch64 architecture profile, G1.3.7
+ *
+ * ID_AA64MMFR0_MSA_FRAC, bits[55:52]
+ * ID_AA64MMFR0_MSA, bits [51:48]
+ */
+#define ID_AA64MMFR0		S3_0_c0_c7_0
+#define ID_AA64MMFR0_MSA_msk	(0xFFUL << 48U)
+#define ID_AA64MMFR0_PMSA_EN	(0x1FUL << 48U)
+
+/*
+ * Global status variable holding the number of HW MPU region indices, which
+ * have been reserved by the MPU driver to program the static (fixed) memory
+ * regions.
+ */
+static uint8_t static_regions_num;
+
+/* Get the number of supported MPU regions. */
+static inline uint8_t get_num_regions(void)
+{
+	uint64_t type;
+
+	type = read_sysreg(MPUIR_EL1);
+	type = type & MPU_IR_REGION_Msk;
+
+	return (uint8_t)type;
+}
+
+/* ARM Core MPU Driver API Implementation for ARM MPU */
+
+/**
+ * @brief enable the MPU
+ */
+void arm_core_mpu_enable(void)
+{
+	uint64_t val;
+
+	val = read_sysreg(sctlr_el1);
+	val |= SCTLR_M_BIT;
+	write_sysreg(val, sctlr_el1);
+	dsb();
+	isb();
+}
+
+/**
+ * @brief disable the MPU
+ */
+void arm_core_mpu_disable(void)
+{
+	uint64_t val;
+
+	/* Force any outstanding transfers to complete before disabling MPU */
+	dmb();
+
+	val = read_sysreg(sctlr_el1);
+	val &= ~SCTLR_M_BIT;
+	write_sysreg(val, sctlr_el1);
+	dsb();
+	isb();
+}
+
+/* ARM MPU Driver Initial Setup
+ *
+ * Configure the cache-ability attributes for all the
+ * different types of memory regions.
+ */
+static void mpu_init(void)
+{
+	/* Device region(s): Attribute-0
+	 * Flash region(s): Attribute-1
+	 * SRAM region(s): Attribute-2
+	 * SRAM no cache-able regions(s): Attribute-3
+	 */
+	uint64_t mair = MPU_MAIR_ATTRS;
+
+	write_sysreg(mair, mair_el1);
```

Comment:
same here: `read_mair_el1()` and `write_mair_el1()`

Reply:
OK, I will